### PR TITLE
Resolves #241 by re-working the way saved subscriptions are handled

### LINKED ISSUE #241: test broker, messages for saved subscriptions
Messages retained for saved subscriptions don't save the v5 properties.
```
        {
            // For each saved subscription, add this message to
            // the list to be sent out when a connection resumes
            // a lost session.
            auto const& idx = subsessions_.get<tag_topic>();
            auto r = idx.equal_range(*topic);
            for (; r.first != r.second; ++r.first) {
                r.first->s->data.emplace_back(
                    topic,
                    contents,
                    std::min(r.first->qos, qos)
                );
            }
        }
```

### PR DESCRIPTION
**Pull request by jonesmz**

### CODE REVIEW

**Comment by jonesmz** [comment]
I have not forgotten about this. I'm just very busy recently.

**Comment by redboltz** [comment]
Don't worry. I understand your situation and I always respect your contribution :)

**Comment by jonesmz** [comment]
This code is now ready for review and merge.

File: test/test_broker.hpp [hunk 1]
```diff
@@ -796,77 +819,156 @@ class test_broker {
...
+                mqtt::make_lambda_visitor<void>(
+                    [&](auto& con) {
+                        con.reset(); // clear the shared pointer, so it doesn't stay alive after this funciton ends.
+                    }
+                ),
+                state.con
+            );
+            act_sess_idx.erase(act_sess_it);
+            BOOST_ASSERT(non_active_sessions_.get<tag_client_id>().find(state.client_id) == non_active_sessions_.get<tag_client_id>().end());
+            auto const& non_active_sess_it = non_active_sessions_.emplace_hint(non_active_sessions_.end(), std::move(state));
```
Comment (redboltz) [inline test/test_broker.hpp]: Passing `end()` to `emplace_hint()` is not efficient. You need to use something like `lower_bound()` and value comparison.

See https://stackoverflow.com/a/41507893/1922763

Comment (redboltz) [reply test/test_broker.hpp]: I noticed that it is from my code ;), but it is confusing. Please replace simple `emplace()` or lower_bound approach. *1

File: test/test_broker.hpp [hunk 2]
```diff
@@ -770,9 +793,9 @@ class test_broker {
                 retains_.erase(*topic);
             }
             else {
-                auto it = retains_.find(*topic);
+                auto const& it = retains_.find(*topic);
                 if(it == retains_.end()) {
-                    retains_.emplace_hint(it, topic, contents, std::move(props), qos);
+                    retains_.emplace_hint(retains_.end(), topic, contents, std::move(props), qos);
```
Comment (redboltz) [inline test/test_broker.hpp]: same as *1

File: test/test_broker.hpp [hunk 3]
```diff
@@ -402,39 +419,134 @@ class test_broker {
...
+                session_state state;
+                non_act_sess_idx.modify(non_act_sess_it,
+                                        [&](session_state & val)
+                                        {
+                                            state = std::move(val);
+                                        });
+                state.con = spep;
+                non_act_sess_idx.erase(non_act_sess_it);
+                act_sess_it = active_sessions_.emplace_hint(active_sessions_.end(), std::move(state));
+                BOOST_ASSERT(act_sess_it->client_id == client_id);
```
Comment (redboltz) [inline test/test_broker.hpp]: same as *1

File: test/test_broker.hpp [hunk 4]
```diff
@@ -402,39 +419,134 @@ class test_broker {
...
+                BOOST_ASSERT(act_sess_it == act_sess_idx.find(client_id));
+            }
+            else {
+                // No existing sessions (active or otherwise) with this client ID
+                // boost::muilti_index should know that emplace_hint with
+                // iterator 'end()' means to skip the internal search
+                act_sess_it = active_sessions_.emplace_hint(active_sessions_.end(), client_id, spep, std::move(will));
```
Comment (redboltz) [inline test/test_broker.hpp]: same as *1

**Comment by redboltz** [comment]
It seems that travis-ci with address sanitizer reports errors.
Could you check this?

NOTE: I fixed and merged memory related issue you reported #300. If it caused the problem, please rebase the PR.

**Comment by jonesmz** [comment]
Do you see any indication of what I'm doing wrong?

I've been trying to find the problem for several days, and I'm not having any luck.

**Comment by redboltz** [comment]
I merged #231, then some conflicts happens. Could you solve it?

**Comment by jonesmz** [comment]
Yes

**Comment by jonesmz** [comment]
Merged

**Comment by redboltz** [comment]
Thank you!

**Pull request merged**